### PR TITLE
Site: editor: Add view site link to site editor nav

### LINKED ISSUE #50405: Add action to "View Site" from the Site Editor sidebar
It is not possible to view your site within the Site View of the Site Editor. You can only do so after navigating into the Canvas and selecting the "View" selector in the editor toolbar.  

Let's explore adding a "View Site" action from the Site Hub perhaps, to enable people to quickly view their site in a new tab. 

### Proposal

CleanShot 2023-05-05 at 15 55 45

### PR DESCRIPTION
**Pull request by glendaviesnz**
## What?
Adds a `View site` link to the site title at top of site editor left nav

## Why?
Fixes: #50405

## How?
Adds an `ExternalLink` component with `getUnstableBase().home` as the url.

## Testing Instructions

- Open site editor and check that no external link icon is visible at top right of left panel
- Hover over  top of left panel and make sure external link icon appears and the `View site` tooltip appears if icon is moused over
- Click site name and check site home opens in new tab

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/3629020/99691876-c373-464d-ac0f-3748f4f6beca

### CODE REVIEW

**Comment by youknowriad** [comment]
Hey @glendaviesnz I've just found this PR and it seems there's some disagreements/confusion about what needs to be done, that we need to clear out.

Separately I've started this PR #50369 as a starting point to address #50378 which also includes a "view post/view page" links. I don't know if we're supposed to have both of these? Anyway, just wanted to link to these and loop in designers to get more clarity @jameskoster @jasmussen @SaxonF

**Comment by jasmussen** [comment]
Riad there's a bit more context in https://github.com/WordPress/gutenberg/issues/50405#issuecomment-1537822246 for this, cc: @richtabor

**Comment by richtabor** [comment]
> Separately I've started this PR https://github.com/WordPress/gutenberg/pull/50369 as a starting point to address https://github.com/WordPress/gutenberg/issues/50378 which also includes a "view post/view page" links. I don't know if we're supposed to have both of these?

The issue is to add a view site link outside of the Edit View, so you don't have to navigate first into a template/page to then view the site.

**Comment by jameskoster** [comment]
The main question for me is whether the link should always open the homepage, or be contextual and open whatever is visible in the frame (assuming it's compatible).

If it's a generic "open the homepage" link, then the placement seems good. Though I'd echo Joen's comment about making use of the site title rather than adding another icon button.

If the link is to be contextually related to the frame then I would question the placement a bit, and suggest exploring something more connected to the frame.

**Comment by richtabor** [comment]
> The main question for me is whether the link should always open the homepage, or be contextual and open whatever is visible in the frame (assuming it's compatible).

I'm thinking homepage.

**Comment by jameskoster** [comment]
Okay.

If we made use of the site title, we could still use the icon attached in a smaller variant:

Screenshot 2023-05-10 at 09 57 36

We use a similar pattern elsewhere:

Screenshot 2023-05-10 at 09 59 46

Or maybe a simple `text-decoration: underline;` would be adequate?

**Comment by jasmussen** [comment]
I would think the underline is sufficient, but either way can we surface the icon on hover only?

**Comment by jameskoster** [comment]
Underline works for me, and avoids the inconsistent icon sizing which looks a bit awkward.

**Comment by jasmussen** [comment]
Can even combine with tooltip if that's not enough. Lots of options!

**Comment by glendaviesnz** [comment]
I have updated it to an underlined link with tooltip:

https://github.com/WordPress/gutenberg/assets/3629020/241ca989-8719-4c97-b3d8-170dd1326c4d

**Comment by richtabor** [comment]
Tooltip, no underline, is my vote 😅

**Comment by jasmussen** [comment]
I wouldn't mind the underline on _hover_, but agree the resting state should not have an underline. Thanks for keeping at this!

**Comment by glendaviesnz** [comment]
> I wouldn't mind the underline on _hover_, but agree the resting state should not have an underline. 

Done 😃

**Comment by jameskoster** [comment]
It was mentioned yesterday that using the site title can create confusion whereby you mistakenly interpret the icon + title to be a single element, which can result in mis-clicks.

An alternative option which avoids adding noise would be to only expose the externalink when you hover the site hub:

view site

I suppose it would need to be permanently visible on mobile, but that seems to be a reasonable trade-off. What do y'all think?

**Comment by richtabor** [comment]
> An alternative option which avoids adding noise would be to only expose the externalink when you hover the site hub

I don't mind this.

**Comment by glendaviesnz** [comment]
> An alternative option which avoids adding noise would be to only expose the externalink when you hover the site hub

>> I don't mind this.

@jasmussen are you also happy with this approach?

**Comment by jasmussen** [comment]
Sure let's go with that.

I still feel like we kind of urgently need a better external link icon, it's huge. Here's a sketch that updates the View menu as well:

Screenshot 2023-05-15 at 10 27 56

That's not blocking for now, just a design task.

**Comment by glendaviesnz** [comment]
I have switched it to the hover only icon:

https://github.com/WordPress/gutenberg/assets/3629020/99691876-c373-464d-ac0f-3748f4f6beca

@jasmussen 18px x 18px seems to balance reasonably well against the command center search icon, but let me know what sort of sizing you think is best.

**Comment by jasmussen** [comment]
Oh my apologies, the icon needs to be updated in the SVG! That's a task for me or someone else, and not something you should do here. Sorry I wasn't clearer. The behavior here is fine even with the oversized icon. I'll make sure I've had enough coffee next time.

**Comment by glendaviesnz** [comment]
> Sorry I wasn't clearer. The behavior here is fine even with the oversized icon. I'll make sure I've had enough coffee next time.

All good, I have actually hard coded the icon to [18px](https://github.com/WordPress/gutenberg/pull/50420/files#diff-6ffc115babb229ea68e4e9dadffa173abc901b2251e1c36d1322d067b6eedb0bR33) in this PR as that is a size that looked ok, so I was just asking if you were happy with that in the meantime while the icon is sorted - so probably more me not being clear 😄

**Comment by jasmussen** [comment]
I'd prefer we always keep the icons 24x24 with very few exceptions. That maintains the spacing as well. I'll be sure to update the ext-link icon when I have a moment.

**Comment by jameskoster** [comment]
+1 it creates a bit of a maintenance headache if we start sizing icons ad hoc. Let's make sure we update the source icon before the next major release, it's on my todo list as well :)

**Comment by glendaviesnz** [comment]
> I'd prefer we always keep the icons 24x24 with very few exceptions.

@jameskoster. @jasmussen if I remove the sizing it defaults to the external link component sizing for the icon which is `1em`:
Screenshot 2023-05-18 at 10 32 23 AM

I can override this with a height and width setting of `initial` which takes it back to 24px, which I have done and pushed. Let me know if any issues with that.

**Comment by jasmussen** [comment]
I'm curious where that 1em fallback value comes from, it should definitely default to 24x24. @mirka in case you have bandwidth, do you know?

**Comment by jasmussen** [comment]
This is feeling good to me, however a few things. The metrics are off in the existing site hub already, with 36px buttons instead of 32px buttons, but nevertheless, the search icon is wrapped in a button:
Screenshot 2023-05-18 at 09 30 08

Should the external link be as well? Not just for the sizing and componentry, but for semantics? 
Screenshot 2023-05-18 at 09 30 19

That's a question, I'm not sure, but would be good to be a bit systematic about the approach there.

Secondly, should we add a simple fade-in/out of the icon when you are hovering? I could help with this on the CSS front, but I'm not sure if we'e started moving towards framer or otherwise for these pieces. [Jay's mockup here](https://github.com/WordPress/gutenberg/pull/50420#issuecomment-1545691128) looks pretty good to me as far as the speed.

**Comment by jasmussen** [comment]
I just created https://github.com/WordPress/gutenberg/pull/50728 to update the default footprint of the external link icon, and in doing so I think I found out [why the 1em values exist](https://github.com/WordPress/gutenberg/pull/50420#issuecomment-1552635458). I think it's because in many cases, the extlink icon is used inside the text of hyperlinks to indicate it as external. Here's an example:

Screenshot 2023-05-18 at 09 48 50

While my PR suggest that a change to 1.5em with the updated footprint can work, I wonder if we shouldn't extract a version of the external link icon from the general icon library, and have it live as a sub-component inside the external link component? Basically we could custom design it to work better in scaling contexts like that, whereas the default icon is designed to always be shown at 24x24. Let me know what you think, happy to work on this, and to be clear, if this is the source of the 1em, it should not block this PR. We probably just need a fade, and then we can land this one.

**Comment by jameskoster** [comment]
> That's a question, I'm not sure, but would be good to be a bit systematic about the approach there

FWIW, in the mockup(s) the _buttons_ are 32px, with 24px _icons_. The footprint matches the site icon for alignment/consistency, while the smaller visual detail is hierarchically appropriate. 

Looking at Storybook, that configuration doesn't seem to be available for icon buttons 🙃 I guess we should make both buttons (command palette and view site) 36 or 24, and address the sizing in a follow-up?

**Comment by jasmussen** [comment]
> FWIW, in the mockup(s) the buttons are 32px, with 24px icons. The footprint matches the site icon for alignment/consistency, while the smaller visual detail is hierarchically appropriate.

Yes! Sorry for the possible confusion there.

> Looking at Storybook, that configuration doesn't seem to be available for icon buttons 🙃 I guess we should make both buttons (command palette and view site) 36 or 24, and address the sizing in a follow-up?

Yep, definitely use the default sizes. The 32px should come separately and holistically as part of #46734.

**Comment by glendaviesnz** [comment]
@jasmussen I have added a basic fade just with CSS, feel free to push changes to this if you are not happy with the timing, etc. At a quick glance it doesn't seem like the components animation stuff covers simple fadein/out yet, but I may be missing something.

> Should the external link be a button as well? Not just for the sizing and componentry, but for semantics?

Semantically it seems correct to have the view link as an `a` as it is just a standard link to open a completely different page in a new tab. The command centre button makes sense as this is an action button that is affecting something in the interface, not linking to somewhere else. But I don't have a strong opinion either way - or perhaps more correctly I do have a strong opinion that it is semantically correct how it is, but that opinion is loosely held so open to changing it 😄 .

**Review by jasmussen — APPROVED** [review]
Thanks for all the work. Fade looks good. 

There's one thing we need to fix, the button should stay fully opaque when focused, otherwise you get this:

Screenshot 2023-05-19 at 08 51 53

With that, I think we can land this as it's a solid improvement that works well.

We do want to follow up, perhaps in the polish phase or as part of components work, to ensure the footprint and spacing is right. A 24x button next to a 36px button is a bit of a mess, semantics aside:
Screenshot 2023-05-19 at 08 52 31

But since we still have #46734 on deck, perhaps that can be separate.

**Comment by jasmussen** [comment]
@ciampo I know you have much on your table. I wonder, should we create a GH board to track all these small component issues that aren't necessarily blocking, but would be good to come back to at some point?

**Comment by carolinan** [comment]
I think this needs accessibility feedback, because having a link that this somewhat important but hidden until you hover makes it very difficult to discover.

When navigating with keyboard only, the icon is not visible when the link is focused, and the tooltip is floating:
Site Editor screen with the tool top visible in the top right corner of the Navigation Sidebar, to the right of the 'go back to the dashboard'

When testing with a screen reader, nothing but "visited link" is announced, and that is not enough information.

**Comment by glendaviesnz** [comment]
@carolinan thanks for looking at this from the accessibility perspective. I will hold off merging until that is discussed.

**Comment by jameskoster** [comment]
Could the link become visible when focus is inside the site hub?

**Review by mirka — CHANGES_REQUESTED** [review]
The `ExternalLink` component is not meant to be used as an icon button like this, at least in its current implementation. Aside from the styling issues, it won't be accessible if we omit `children` and don't provide an aria-label.

Can we use a [`<Button icon={} href={} label={} target="_blank" >`](https://wordpress.github.io/gutenberg/?path=/story/components-button--icon) instead? I think this would resolve the accessibility and styling issues.

(And if this kind of thing is required more often, we could then consider extracting some of the `ExternalLink` logic and reusing it in `Button`.)

**Comment by carolinan** [comment]
Buttons should only be used for actions, not for linking. Are these the only two options?

**Comment by mirka** [comment]
> Buttons should only be used for actions, not for linking. Are these the only two options?

The `Button` component will automatically render as a `<a>` when there is a `href` prop, so it will be semantically correct 👍

**Comment by glendaviesnz** [comment]
I have switched to button and also set the icon to display if the hub content area is focused, but I had trouble getting that to focus using the keyboard.

**Comment by jameskoster** [comment]
A couple of details (apologies, you must be tired of this one):

* The button disappears when you hover the search button which feels a bit unexpected. Ideally it's visible when the cursor is inside `edit-site-site-hub`.
* The space between external / search buttons is 8px which is a touch large. If we can reduce that to 4px it looks a little neater.

Aside from that and any outstanding a11y issues we should be good to merge.

**Comment by ciampo** [comment]
> @ciampo I know you have much on your table. I wonder, should we create a GH board to track all these small component issues that aren't necessarily blocking, but would be good to come back to at some point?

Hey @jasmussen , that sounds good to me. We won't necessarily have the capacity to work on these issues firsthand, but we're always happy to help with reviews and advice (and with some authoring when necessary like we did on the [editor frame resizing PR](https://github.com/WordPress/gutenberg/pull/49910))

File: packages/edit-site/src/components/site-hub/style.scss
Comment (mirka) [inline packages/edit-site/src/components/site-hub/style.scss]: > I had trouble getting that to focus using the keyboard

The focus style is on the wrong element:

```diff
diff --git a/packages/edit-site/src/components/site-hub/style.scss b/packages/edit-site/src/components/site-hub/style.scss
index d38ca4068b..e8997fb37c 100644
--- a/packages/edit-site/src/components/site-hub/style.scss
+++ b/packages/edit-site/src/components/site-hub/style.scss
@@ -28,6 +28,7 @@
 		&:focus {
 			outline: none;
 			box-shadow: none;
+			opacity: 1;
 		}
 		svg {
 			fill: $white;
@@ -38,11 +39,6 @@
 			opacity: 1;
 		}
 	}
-	&:focus {
-		.edit-site-site-hub__site-view-link {
-			opacity: 1;
-		}
-	}
 }
 
 .edit-site-site-hub__title {
```

**Review by mirka — APPROVED** [review]
Good to go on my end, once the focus styles are fixed 🚀

File: packages/edit-site/src/components/site-hub/style.scss
Comment (glendaviesnz) [reply packages/edit-site/src/components/site-hub/style.scss]: Thanks, fixed.

**Comment by glendaviesnz** [comment]
@jameskoster I have fixed the hover issue and reduced the gap.

**Comment by jameskoster** [comment]
Thanks Glen! Pending a11y sign-off this is looking good to go.

**Comment by mirka** [comment]
Accessibility-wise:

- Keyboard navigable ✅
- Button (actually an `<a>`) semanticallly correct and labeled ✅
- We should probably [append a warning](https://www.w3.org/TR/WCAG20-TECHS/H83.html) that the link opens in a new tab. Does the Button component allow different content in the tooltip and the aria label? So like `label="View site"` (for tooltip) and `aria-label="View site (opens in a new tab)"` (for SRs only).

**Comment by glendaviesnz** [comment]
> * Does the Button component allow different content in the tooltip and the aria label? So like `label="View site"` (for tooltip) and `aria-label="View site (opens in a new tab)"` (for SRs only).

Yes, that seems to work, have [added that](https://github.com/WordPress/gutenberg/pull/50420/commits/11f798b22086e1aad77efb6d6702123be7cfb9b0).

**Comment by glendaviesnz** [comment]
I am going to go ahead and merge as it should be easy to follow up on if any other accessibility issues are raised.

**Comment by afercia** [comment]
Just noticed that this PR broke the layout of the Site hub when tabbing with the keyboard. Created https://github.com/WordPress/gutenberg/issues/51257 and suggested a possible fix there.

**Comment by joedolson** [comment]
A couple additional accessibility follow-ups:

1) The focus state for this control doesn't match neighboring controls:

open-command-palette-focus-state

view-site-focus-state

2) The tooltip text doesn't notify the user the link will open in a new tab., it's only available in the aria-label. 

3) What is the reasoning behind having this link hidden by default? This significantly reduces discoverability, and I'm not clear that it provides any advantages.

**Comment by glendaviesnz** [comment]
@joedolson I have added a [PR to fix the focus and label issues](https://github.com/WordPress/gutenberg/pull/52744). @jameskoster would be the person to answer the question about [why it is only visible on mouseover/focus](https://github.com/WordPress/gutenberg/pull/50420#issuecomment-1545691128).

**Comment by jameskoster** [comment]
Iirc, as it is not 100% clear how the 'site hub' will evolve as we explore the [broader admin design](https://make.wordpress.org/core/2023/07/12/admin-design/) there was motivation to keep it as lightweight as possible for now. Nothing is set in stone, and it may be something we revisit later.

**Comment by joedolson** [comment]
We may have different definitions of lightweight; I see the hidden link as a more complex design with a higher cognitive and discovery load. However, I do appreciate that until there's some kind of admin and management component in place, there isn't really a well-defined context for this. But not a fan of hidden controls.

**Comment by afercia** [comment]
> But not a fan of hidden controls.

I'd totally second this, and I'd like to add that in the last 6 years one of the most common accessibility concerns that was very often reported is specifically about UI controls that continuously appear and disappear. That's an accessibility anti-pattern and I'd greatly appreciate all designers to be more aware of that. Thank you.

**Pull request merged**
